src/v/security/oidc_service.cc

```diff
@@ -0,0 +1,249 @@
+    uint16_t port = url.port_number();
+    if (url.port().empty()) {
+        port = is_https ? 443 : 80;
+    }
+
+    constexpr auto sv_to_sv = [](auto sv) -> std::string_view {
+        return {sv.data(), sv.size()};
+    };
+
+    ss::sstring host{url.encoded_host()};
+    std::optional<ss::sstring> tls_host;
+    ss::shared_ptr<ss::tls::certificate_credentials> creds;
+    if (is_https) {
+        tls_host.emplace(host);
+        ss::tls::credentials_builder builder;
+        builder.set_client_auth(ss::tls::client_auth::NONE);
+        co_await builder.set_system_trust();
+        creds = builder.build_certificate_credentials();
+        creds->set_dn_verification_callback([](
+                                              ss::tls::session_type type,
+                                              ss::sstring subject,
+                                              ss::sstring issuer) {
+            vlog(
+              seclog.trace,
+              "type: ?, subject: {}, issuer: {}",
+              (uint8_t)type,
+              subject,
+              issuer);
+        });
+    }
+    http::client client{net::base_transport::configuration{
+      .server_addr = {host, port},
+      .credentials = creds,
+      .tls_sni_hostname = tls_host}};
+
+    http::client::request_header req_hdr;
+    req_hdr.method(boost::beast::http::verb::get);
+    req_hdr.target(sv_to_sv(url.encoded_path()));
+    req_hdr.insert(boost::beast::http::field::host, sv_to_sv(host));
+    req_hdr.insert(boost::beast::http::field::accept, "*/*");
+
+    co_return co_await http::with_client(
+      std::move(client),
+      [req_hdr{std::move(req_hdr)}](
+        auto& client) mutable -> ss::future<ss::sstring> {
+          auto res = co_await client.request(std::move(req_hdr));
+
+          ss::sstring response_body;
+          while (!res->is_done()) {
+              iobuf buf = co_await res->recv_some();
+              for (auto& fragm : buf) {
+                  response_body.append(fragm.get(), fragm.size());
+              }
+          }
+          co_return response_body;
+      });
+}
+} // namespace
+
+struct service::impl {
+    ss::shard_id update_shard_id{0};
+    impl(
+      config::binding<std::vector<ss::sstring>> sasl_mechanisms,
+      config::binding<std::vector<ss::sstring>> http_authentication,
+      config::binding<ss::sstring> discovery_url,
+      config::binding<ss::sstring> token_audience,
+      config::binding<std::chrono::seconds> clock_skew_tolerance)
+      : _verifier{}
+      , _sasl_mechanisms{std::move(sasl_mechanisms)}
+      , _http_authentication{std::move(http_authentication)}
+      , _discovery_url{std::move(discovery_url)}
+      , _token_audience{std::move(token_audience)}
+      , _clock_skew_tolerance{std::move(clock_skew_tolerance)} {
+        _sasl_mechanisms.watch([this]() { return update(); });
+        _http_authentication.watch([this]() { return update(); });
+        _discovery_url.watch([this]() { return update(); });
```

Comment:
>_sasl_mechanisms.watch([this]() { return update(); });

the function signature passed to watch is `void()` -- so this has the affect of backgrounding the future returned from update. i suspect that's not what is intended?

Reply:
It's intended; `update` holds the gate.

Changed to `ssx::spawn_with_gate` to better handle exceptions during shutdown.

Reply:
> Changed to ssx::spawn_with_gate to better handle exceptions during shutdown.

thank you